File: src/main/java/com/drunkensoftball/api/support/jackson/BaseEntitySerializer.java
package com.drunkensoftball.api.support.jackson;

import com.drunkensoftball.api.domain.BaseEntity;
import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.JsonSerializer;
import com.fasterxml.jackson.databind.SerializerProvider;

import java.io.IOException;

public class BaseEntitySerializer<T extends BaseEntity>  extends JsonSerializer<T> {

    protected static final String CREATED = "created";
    protected static final String UPDATED = "updated";
    protected static final String ID = "id";
    protected static final String UUID = "uuid";

    protected static final String DISPAY_NAME = "displayName";
    protected static final String USERNAME = "username";
    protected static final String MANAGED_TEAMS = "managedTeams";
    protected static final String JOINED_TEAMS = "joinedTeams";
    protected static final String USER = "user";
    protected static final String TEAM = "team";
    protected static final String NAME = "name";
    protected static final String MANAGER = "manager";
    protected static final String ROSTER = "roster";
    protected static final String BATTING_POSITION = "battingPosition";
    protected static final String FIELD_POSITION = "fieldPosition";

    @Override
    public void serialize(final T t,
                          final JsonGenerator jsonGenerator,
                          final SerializerProvider provider) throws IOException {

        jsonGenerator.writeStartObject();
        writeAbstractEntityFields(t, jsonGenerator);

        this.handleSubclass(t, jsonGenerator);

        jsonGenerator.writeEndObject();
    }

    protected void handleSubclass(final T t, final JsonGenerator jsonGenerator) throws IOException {
    }

    protected <U extends BaseEntity> void writeAbstractEntityFields(final U u, final JsonGenerator jsonGenerator) throws IOException {
        jsonGenerator.writeStringField(UUID, u.getUuid());
        this.handleCreateUpdateFields(u, jsonGenerator);
    }

    protected <U extends BaseEntity> void handleCreateUpdateFields(final U u, final JsonGenerator jsonGenerator) throws IOException {
        if (null != u.getCreated()) {
            jsonGenerator.writeNumberField(CREATED, u.getCreated().getTimeInMillis());
        }

        if (null != u.getUpdated()) {
            jsonGenerator.writeNumberField(UPDATED, u.getUpdated().getTimeInMillis());
        }
    }
}

File: src/main/kotlin/com/drunkensoftball/api/support/jackson/BaseEntitySerializer.kt
package com.drunkensoftball.api.support.jackson


import com.drunkensoftball.api.domain.BaseEntity
import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.databind.JsonSerializer
import com.fasterxml.jackson.databind.SerializerProvider

import java.io.IOException

open class BaseEntitySerializer<T : BaseEntity> : JsonSerializer<T>() {

    @Throws(IOException::class)
    override fun serialize(entity: T,
                           jsonGenerator: JsonGenerator,
                           provider: SerializerProvider) {

        jsonGenerator.writeStartObject()
        writeAbstractEntityFields(entity, jsonGenerator)

        this.handleSubclass(entity, jsonGenerator)

        jsonGenerator.writeEndObject()
    }

    @Throws(IOException::class)
    fun handleSubclass(t: T, jsonGenerator: JsonGenerator) {
    }

    @Throws(IOException::class)
    fun <U : BaseEntity> writeAbstractEntityFields(u: U, jsonGenerator: JsonGenerator) {
        jsonGenerator.writeStringField(UUID, u.uuid)
        this.handleCreateUpdateFields(u, jsonGenerator)
    }

    @Throws(IOException::class)
    fun <U : BaseEntity> handleCreateUpdateFields(u: U, jsonGenerator: JsonGenerator) {
        if (null != u.created) {
            jsonGenerator.writeNumberField(CREATED, u.created!!.timeInMillis)
        }

        if (null != u.updated) {
            jsonGenerator.writeNumberField(UPDATED, u.updated!!.timeInMillis)
        }
    }

    companion object {

        const val CREATED = "created"
        const val UPDATED = "updated"
        const val ID = "id"
        const val UUID = "uuid"

        const val DISPLAY_NAME = "displayName"
        const val FIRST_NAME = "firstName"
        const val LAST_NAME = "lastName"
        const val USERNAME = "username"
        const val EMAIL = "email"
        const val MANAGED_TEAMS = "managedTeams"
        const val JOINED_TEAMS = "joinedTeams"
        const val USER = "user"
        const val TOKEN = "token"
        const val TEAM = "team"
        const val NAME = "name"
        const val MANAGER = "manager"
        const val ROSTER = "roster"
        const val GAMES = "games"
        const val OPPONENT_NAME = "opponentName"
        const val BATTING_POSITION = "battingPosition"
        const val FIELD_POSITION = "fieldPosition"
    }
}